ipmininet/tests/test_bgp.py

```diff
@@ -115,3 +122,228 @@ def test_bgp_daemon_params(bgp_params, expected_cfg):
+        'fd00:4:1::2'
+    ],
+    'as1r5': [
+        'fd00:1:6::2',
+        'fd00:1:7::1',
+        'fd00:4:2::2'
+    ],
+    'as1r6': [
+        'fd00:1:1::2',
+        'fd00:1:3::2',
+        'fd00:1:7::2',
+        'fd00:5:1::2'
+    ],
+    'as2r1': [
+        'fd00:2:1::2',
+        'fd00:2:2::1',
+        'dead:beef::1'
+    ],
+    'as2h1': [
+        'dead:beef::2'
+    ],
+    'as3r1': [
+        'fd00:3:1::1',
+        'fd00:5:2::1'
+    ],
+    'as5r1': [
+        'fd00:5:1::1',
+        'fd00:5:2::2',
+        'fd00:2:1::1'
+    ],
+    'as4r1': [
+        'fd00:4:2::1',
+        'fd00:2:2::2',
+        'fd00:4:3::1'
+    ], 
+    'as4r2': [
+        'fd00:4:1::1',
+        'fd00:4:3::2'
+    ]
+}
+
+
+local_pref_paths = [
+    ['as1r1', 'as1r6', 'dead:beef::'],
+    ['as1r2', 'as1r3', 'as1r6', 'dead:beef::'],
+    ['as1r3', 'as1r6', 'dead:beef::'],
+    ['as1r4', 'as1r5', 'as1r6', 'dead:beef::'],
+    ['as1r5', 'as1r6', 'dead:beef::'],
+    ['as1r6', 'dead:beef::']
+]
+
+
+@require_root
+def test_bgp_local_pref():
+    try:
+        net = IPNet(topo=BGPTopoLocalPref())
+        net.start()
+        for path in local_pref_paths:
+            assert_path_bgp(net, path, config_ip1)
+        net.stop()
+    finally:
+        cleanup()
+
+
+med_paths = [
+    ['as1r1', 'as1r6', 'as1r5', 'dead:beef::'],
+    ['as1r2', 'as1r4', 'as1r5', 'dead:beef::'],
+    ['as1r3', 'as1r6', 'as1r5', 'dead:beef::'],
+    ['as1r4', 'as1r5', 'dead:beef::'],
+    ['as1r5', 'dead:beef::'],
+    ['as1r6', 'as1r5', 'dead:beef::']
+]
+
+
+@require_root
+def test_bgp_med():
+    try:
+        net = IPNet(topo=BGPTopoMed())
+        net.start()
+        for path in med_paths:
+            assert_path_bgp(net, path, config_ip1)
+        net.stop()
+    finally:
+        cleanup()
+
+
+rr_paths = [
+    ['as1r1', 'as1r6', 'as5r1', 'dead:beef::'],
+    ['as1r2', 'as1r4', 'as4r2', 'as4r1', 'dead:beef::'],
+    # ['as1r3', 'as1r1, 'as1r6', 'as5r1', 'dead:beef::'], to fix
+    ['as1r4', 'as4r2', 'as4r1', 'dead:beef::'],
+    ['as1r5', 'as4r1', 'dead:beef::'],
+    ['as1r6', 'as5r1', 'dead:beef::']
+]
+
+
+@require_root
+def test_bgp_rr():
+    try:
+        net = IPNet(topo=BGPTopoRR())
+        net.start()
+        for path in rr_paths:
+            assert_path_bgp(net, path, config_ip2)
+        net.stop()
+    finally:
+        cleanup()
+
+
+full_paths = [
+    ['as1r1', 'as1r3', 'as1r6', 'dead:beef::'],
+    ['as1r2', 'as1r3', 'as1r6', 'dead:beef::'],
+    ['as1r3', 'as1r6', 'dead:beef::'],
+    ['as1r4', 'as1r2' ,'as1r3', 'as1r6', 'dead:beef::'],
+    ['as1r5', 'as1r6', 'dead:beef::'],
+    ['as1r6', 'dead:beef::']
+]
+
+
+@require_root
+def test_bgp_full_config():
+    try:
+        net = IPNet(topo=BGPTopoFull())
+        net.start()
+        for path in full_paths:
+            assert_path_bgp(net, path, config_ip2)
+        net.stop()
+    finally:
+        cleanup()
+
+
+def assert_path_bgp(net, expected_path, nodes_ips, timeout=300, udp=True):
```

Comment:
- using icmp messages for the traceroute didn't work for reasons we didn't find
- I didn't succeed to get the static addresses of the examples dynamically in assert_path, it's why I created dictionaries with these addresses and a new function using these dictionaries.

Do you prefer putting this function private or removing the static addresses?

Reply:
>     * using icmp messages for the traceroute didn't work for reasons we didn't find

This is likely because you hit some icmp limit. Hosts will stop to answer if they receive too many icmps.
But then, why are you adding the option of using traceroute with icmps?
By default traceroute uses udp and you add the possibility to use icmps in your PR.
Then, you should remove the code that uses the udp parameter.

>     * I didn't succeed to get the static addresses of the examples dynamically in assert_path, it's why I created dictionaries with these addresses and a new function using these dictionaries.

The following code works well but I don't think you need it in your code actually.
For instance, in the test for the local pref topology, doing an assert_path(['as1r1', 'as1r6', 'as4r1', 'as4h1']) should work if I read your example topology correctly.
```python
net["<node-name>"].intf("<interface-name>").ip6
```

> Do you prefer putting this function private or removing the static addresses?

I prefer removing static addresses (see comment above for the reason)

Reply:
> But then, why are you adding the option of using traceroute with icmps?
The function traceroute was originally using the icmps

Reply:
> > But then, why are you adding the option of using traceroute with icmps?
> 
> The function traceroute was originally using the icmps

My bad, I didn't read the diff correctly ^^
Ok then, detail the example that is not working with ICMP that justify the need for adding it to the code base in the message of the commit that adds UDP traceroutes :)